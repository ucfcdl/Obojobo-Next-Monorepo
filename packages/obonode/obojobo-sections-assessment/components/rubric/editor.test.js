/* eslint no-undefined: 0 */

import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import KeyDownUtil from 'obojobo-document-engine/src/scripts/oboeditor/util/keydown-util'
jest.mock('obojobo-document-engine/src/scripts/oboeditor/util/keydown-util')
import SlateReact from 'slate-react'
jest.mock('slate-react')

import Rubric from './editor'
const RUBRIC_NODE = 'ObojoboDraft.Sections.Assessment.Rubric'
const MOD_NODE = 'ObojoboDraft.Sections.Assessment.Rubric.Mod'
const MOD_LIST_NODE = 'ObojoboDraft.Sections.Assessment.Rubric.ModList'

describe('Rubric editor', () => {
	test('Node component', () => {
		const Node = Rubric.components.Node
		const component = renderer.create(
			<Node
				node={{
					data: {
						get: () => {
							return {}
						}
					}
				}}
			/>
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Node component adds child', () => {
		jest.spyOn(window, 'prompt')
		window.prompt.mockReturnValueOnce(null)

		const editor = {
			insertNodeByKey: jest.fn()
		}

		const Node = Rubric.components.Node
		const component = mount(
			<Node
				node={{
					data: {
						get: () => {
							return {}
						}
					},
					nodes: { size: 0 }
				}}
				editor={editor}
			/>
		)
		const tree = component.html()

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(tree).toMatchSnapshot()
	})

	test('Node component adds child to existing mod list', () => {
		jest.spyOn(window, 'prompt')
		window.prompt.mockReturnValueOnce(null)

		const editor = {
			insertNodeByKey: jest.fn()
		}

		const Node = Rubric.components.Node
		const component = mount(
			<Node
				node={{
					data: {
						get: () => {
							return {}
						}
					},
					nodes: {
						size: 5,
						get: jest.fn().mockReturnValueOnce({
							key: 'mockModList',
							nodes: { size: 0 }
						})
					}
				}}
				editor={editor}
			/>
		)
		const tree = component.html()

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(tree).toMatchSnapshot()
	})

	test('Rubric component deletes self', () => {
		const editor = {
			removeNodeByKey: jest.fn()
		}

		const Node = Rubric.components.Node
		const component = mount(
			<Node
				node={{
					data: {
						get: () => {
							return {}
						}
					}
				}}
				editor={editor}
			/>
		)
		const tree = component.html()

		component
			.find('button')
			.at(1)
			.simulate('click')

		expect(tree).toMatchSnapshot()
		expect(editor.removeNodeByKey).toHaveBeenCalled()
	})

	test('ModList component', () => {
		const Node = Rubric.components.ModList
		const component = renderer.create(
			<Node
				node={{
					data: {
						get: () => {
							return {}
						}
					}
				}}
			/>
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Mod component', () => {
		const Node = Rubric.components.Mod
		const component = renderer.create(<Node />)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Mod component deletes self', () => {
		const editor = {
			removeNodeByKey: jest.fn(),
			value: {
				document: {
					getDescendant: jest.fn().mockReturnValueOnce({
						nodes: {
							get: jest.fn().mockReturnValueOnce(true)
						}
					})
				}
			}
		}

		const Node = Rubric.components.Mod
		const component = mount(
			<Node
				node={{
					data: {
						get: () => {
							return {}
						}
					}
				}}
				editor={editor}
				parent={{ key: 'mockParent' }}
			/>
		)
		const tree = component.html()

		component.find('button').simulate('click')

		expect(tree).toMatchSnapshot()
		expect(editor.removeNodeByKey).toHaveBeenCalled()
	})

	test('Mod component deletes self and parent', () => {
		const editor = {
			removeNodeByKey: jest.fn(),
			value: {
				document: {
					getDescendant: jest.fn().mockReturnValueOnce({
						nodes: {
							get: jest.fn() // No sibling
						}
					})
				}
			}
		}

		const Node = Rubric.components.Mod
		const component = mount(
			<Node
				node={{
					data: {
						get: () => {
							return {}
						}
					}
				}}
				editor={editor}
				parent={{ key: 'mockParent' }}
			/>
		)
		const tree = component.html()

		component.find('button').simulate('click')

		expect(tree).toMatchSnapshot()
		expect(editor.removeNodeByKey).toHaveBeenCalled()
	})

	test('slateToObo converts a Slate node to an OboNode with content', () => {
		const slateNode = {
			key: 'mockKey',
			type: 'mockType',
			data: {
				get: () => {
					return null
				}
			},
			nodes: [
				{
					data: {
						get: () => {
							return {}
						}
					}
				},
				{
					data: {
						get: () => {
							return {}
						}
					},
					text: ''
				},
				{
					type: MOD_LIST_NODE,
					data: {
						get: () => {
							return {}
						}
					},
					nodes: [
						{
							key: 'mockMod',
							nodes: {
								get: () => {
									return { text: 'mockParameter' }
								}
							}
						}
					]
				}
			]
		}
		const oboNode = Rubric.helpers.slateToObo(slateNode)

		expect(oboNode).toMatchSnapshot()
	})

	test('slateToObo converts a Slate node to an OboNode with no mods', () => {
		const slateNode = {
			key: 'mockKey',
			type: 'mockType',
			data: {
				get: () => {
					return null
				}
			},
			nodes: []
		}
		const oboNode = Rubric.helpers.slateToObo(slateNode)

		expect(oboNode).toMatchSnapshot()
	})

	test('oboToSlate converts an OboComponent to a Slate node', () => {
		const oboNode = {
			passingAttemptScore: 0,
			passedResult: 100,
			failedResult: 0,
			unableToPassResult: 0,
			mods: [
				{
					attemptCondition: 1,
					reward: 10
				}
			]
		}
		const slateNode = Rubric.helpers.oboToSlate(oboNode)

		expect(slateNode).toMatchSnapshot()
	})

	test('oboToSlate converts an OboComponent to a Slate node without mods', () => {
		const oboNode = {
			passingAttemptScore: 0,
			passedResult: 100,
			failedResult: 0,
			unableToPassResult: 0
		}
		const slateNode = Rubric.helpers.oboToSlate(oboNode)

		expect(slateNode).toMatchSnapshot()
	})

	test('plugins.onPaste deals with no rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: () => false
				}
			}
		}
		editor.insertBlock = jest.fn().mockReturnValueOnce(editor)

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onPaste(event, editor, jest.fn())

		expect(event.preventDefault).not.toHaveBeenCalled()
	})

	test('plugins.onPaste deals with pasting into rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: (key, funct) => {
						funct({ type: 'mockType' })
						return true
					}
				}
			}
		}
		editor.insertText = jest.fn().mockReturnValueOnce(editor)

		SlateReact.getEventTransfer.mockReturnValueOnce({ text: 'mock text' })

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onPaste(event, editor, jest.fn())

		expect(editor.insertText).toHaveBeenCalled()
	})

	test('plugins.onCut deals with no rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: () => false
				}
			},
			extractTextToFragment: jest.fn()
		}
		editor.insertBlock = jest.fn().mockReturnValueOnce(editor)

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onCut(event, editor, jest.fn())

		expect(editor.extractTextToFragment).not.toHaveBeenCalled()
		expect(KeyDownUtil.deleteNodeContents).not.toHaveBeenCalled()
	})

	test('plugins.onCut deals with cutting from rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: (key, funct) => {
						funct({ type: 'mockType' })
						return true
					}
				},
				fragment: { text: 'selected text' }
			},
			extractTextToFragment: jest.fn()
		}

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onCut(event, editor, jest.fn())

		expect(editor.extractTextToFragment).toHaveBeenCalled()
		expect(KeyDownUtil.deleteNodeContents).toHaveBeenCalled()
	})

	test('plugins.onCopy deals with no rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: () => false
				}
			},
			extractTextToFragment: jest.fn()
		}

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onCopy(event, editor, jest.fn())

		expect(editor.extractTextToFragment).not.toHaveBeenCalled()
	})

	test('plugins.onCopy deals with copying from rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: (key, funct) => {
						funct({ type: 'mockType' })
						return true
					}
				},
				fragment: { text: 'selected text' }
			},
			extractTextToFragment: jest.fn()
		}

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onCopy(event, editor, jest.fn())

		expect(editor.extractTextToFragment).toHaveBeenCalled()
	})

	test('plugins.renderNode renders the rubric when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: RUBRIC_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Rubric.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode renders a modlist when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: MOD_LIST_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Rubric.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode renders a modlist when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: MOD_LIST_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Rubric.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode renders a modlist when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: MOD_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Rubric.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode calls next', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: 'mockNode',
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		const next = jest.fn()

		expect(Rubric.plugins.renderNode(props, null, next)).toMatchSnapshot()
		expect(next).toHaveBeenCalled()
	})

	test('extractTextToFragment copies the current text', () => {
		const editor = {
			value: {
				fragment: { text: 'selected text' }
			}
		}
		expect(Rubric.plugins.queries.extractTextToFragment(editor)).toMatchSnapshot()
	})
})
