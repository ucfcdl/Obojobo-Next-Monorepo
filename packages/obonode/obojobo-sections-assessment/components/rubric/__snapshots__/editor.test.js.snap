// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rubric editor Mod component 1`] = `
<div
  className="mod pad"
>
  <div
    className="obojobo-draft--components--button is-not-dangerous align-center delete-button"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      ×
    </button>
  </div>
</div>
`;

exports[`Rubric editor Mod component deletes self 1`] = `"<div class=\\"mod pad\\"><div class=\\"obojobo-draft--components--button is-not-dangerous align-center delete-button\\"><button class=\\"button\\">×</button></div></div>"`;

exports[`Rubric editor Mod component deletes self and parent 1`] = `"<div class=\\"mod pad\\"><div class=\\"obojobo-draft--components--button is-not-dangerous align-center delete-button\\"><button class=\\"button\\">×</button></div></div>"`;

exports[`Rubric editor ModList component 1`] = `
<div>
  <p
    contentEditable={false}
  >
    Mods:
  </p>
</div>
`;

exports[`Rubric editor Node component 1`] = `
<div
  className="rubric pad"
>
  <h1
    contentEditable={false}
  >
    Rubric
  </h1>
  <div
    className="parameter-node"
    contentEditable={false}
  >
    Type: undefined
  </div>
  <div
    className="obojobo-draft--components--button is-not-dangerous align-center add-button"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Add Mod
    </button>
  </div>
  <div
    className="obojobo-draft--components--button is-not-dangerous align-center delete-button"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      ×
    </button>
  </div>
</div>
`;

exports[`Rubric editor Node component adds child 1`] = `"<div class=\\"rubric pad\\"><h1 contenteditable=\\"false\\">Rubric</h1><div class=\\"parameter-node\\" contenteditable=\\"false\\">Type: undefined</div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center add-button\\"><button class=\\"button\\">Add Mod</button></div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center delete-button\\"><button class=\\"button\\">×</button></div></div>"`;

exports[`Rubric editor Node component adds child to existing mod list 1`] = `"<div class=\\"rubric pad\\"><h1 contenteditable=\\"false\\">Rubric</h1><div class=\\"parameter-node\\" contenteditable=\\"false\\">Type: undefined</div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center add-button\\"><button class=\\"button\\">Add Mod</button></div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center delete-button\\"><button class=\\"button\\">×</button></div></div>"`;

exports[`Rubric editor Rubric component deletes self 1`] = `"<div class=\\"rubric pad\\"><h1 contenteditable=\\"false\\">Rubric</h1><div class=\\"parameter-node\\" contenteditable=\\"false\\">Type: undefined</div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center add-button\\"><button class=\\"button\\">Add Mod</button></div><div class=\\"obojobo-draft--components--button is-not-dangerous align-center delete-button\\"><button class=\\"button\\">×</button></div></div>"`;

exports[`Rubric editor extractTextToFragment copies the current text 1`] = `
Immutable.Record {
  "data": Immutable.Map {},
  "key": "11",
  "nodes": Immutable.List [
    Immutable.Record {
      "data": Immutable.Map {},
      "key": "12",
      "nodes": Immutable.List [
        Immutable.Record {
          "data": Immutable.Map {},
          "key": "13",
          "nodes": Immutable.List [
            Immutable.Record {
              "data": Immutable.Map {
                "indent": 0,
              },
              "key": "14",
              "nodes": Immutable.List [
                Immutable.Record {
                  "leaves": Immutable.List [
                    Immutable.Record {
                      "marks": Immutable.Set [],
                      "text": "selected text",
                    },
                  ],
                  "key": "15",
                },
              ],
              "type": "ObojoboDraft.Chunks.Text.TextLine",
            },
          ],
          "type": "ObojoboDraft.Chunks.Text",
        },
      ],
      "type": "oboeditor.component",
    },
  ],
}
`;

exports[`Rubric editor oboToSlate converts an OboComponent to a Slate node 1`] = `
Object {
  "data": Object {
    "content": Object {
      "failedResult": 0,
      "mods": Array [
        Object {
          "attemptCondition": 1,
          "reward": 10,
        },
      ],
      "passedResult": 100,
      "passingAttemptScore": 0,
      "unableToPassResult": 0,
    },
  },
  "nodes": Array [
    Object {
      "data": Object {
        "display": "Passing Score",
        "name": "passingAttemptScore",
      },
      "nodes": Array [
        Object {
          "leaves": Array [
            Object {
              "text": 0,
            },
          ],
          "object": "text",
        },
      ],
      "object": "block",
      "type": "oboeditor.text-parameter",
    },
    Object {
      "data": Object {
        "display": "Passed Result",
        "name": "passedResult",
      },
      "nodes": Array [
        Object {
          "leaves": Array [
            Object {
              "text": 100,
            },
          ],
          "object": "text",
        },
      ],
      "object": "block",
      "type": "oboeditor.text-parameter",
    },
    Object {
      "data": Object {
        "display": "Failed Result",
        "name": "failedResult",
      },
      "nodes": Array [
        Object {
          "leaves": Array [
            Object {
              "text": 0,
            },
          ],
          "object": "text",
        },
      ],
      "object": "block",
      "type": "oboeditor.text-parameter",
    },
    Object {
      "data": Object {
        "display": "Unable to Pass Result",
        "name": "unableToPassResult",
      },
      "nodes": Array [
        Object {
          "leaves": Array [
            Object {
              "text": 0,
            },
          ],
          "object": "text",
        },
      ],
      "object": "block",
      "type": "oboeditor.text-parameter",
    },
    Object {
      "nodes": Array [
        Object {
          "nodes": Array [
            Object {
              "data": Object {
                "display": "Attempt Condition",
                "name": "attemptCondition",
              },
              "nodes": Array [
                Object {
                  "leaves": Array [
                    Object {
                      "text": 1,
                    },
                  ],
                  "object": "text",
                },
              ],
              "object": "block",
              "type": "oboeditor.text-parameter",
            },
            Object {
              "data": Object {
                "display": "Reward",
                "name": "reward",
              },
              "nodes": Array [
                Object {
                  "leaves": Array [
                    Object {
                      "text": 10,
                    },
                  ],
                  "object": "text",
                },
              ],
              "object": "block",
              "type": "oboeditor.text-parameter",
            },
          ],
          "object": "block",
          "type": "ObojoboDraft.Sections.Assessment.Rubric.Mod",
        },
      ],
      "object": "block",
      "type": "ObojoboDraft.Sections.Assessment.Rubric.ModList",
    },
  ],
  "object": "block",
  "type": "ObojoboDraft.Sections.Assessment.Rubric",
}
`;

exports[`Rubric editor oboToSlate converts an OboComponent to a Slate node without mods 1`] = `
Object {
  "data": Object {
    "content": Object {
      "failedResult": 0,
      "passedResult": 100,
      "passingAttemptScore": 0,
      "unableToPassResult": 0,
    },
  },
  "nodes": Array [
    Object {
      "data": Object {
        "display": "Passing Score",
        "name": "passingAttemptScore",
      },
      "nodes": Array [
        Object {
          "leaves": Array [
            Object {
              "text": 0,
            },
          ],
          "object": "text",
        },
      ],
      "object": "block",
      "type": "oboeditor.text-parameter",
    },
    Object {
      "data": Object {
        "display": "Passed Result",
        "name": "passedResult",
      },
      "nodes": Array [
        Object {
          "leaves": Array [
            Object {
              "text": 100,
            },
          ],
          "object": "text",
        },
      ],
      "object": "block",
      "type": "oboeditor.text-parameter",
    },
    Object {
      "data": Object {
        "display": "Failed Result",
        "name": "failedResult",
      },
      "nodes": Array [
        Object {
          "leaves": Array [
            Object {
              "text": 0,
            },
          ],
          "object": "text",
        },
      ],
      "object": "block",
      "type": "oboeditor.text-parameter",
    },
    Object {
      "data": Object {
        "display": "Unable to Pass Result",
        "name": "unableToPassResult",
      },
      "nodes": Array [
        Object {
          "leaves": Array [
            Object {
              "text": 0,
            },
          ],
          "object": "text",
        },
      ],
      "object": "block",
      "type": "oboeditor.text-parameter",
    },
  ],
  "object": "block",
  "type": "ObojoboDraft.Sections.Assessment.Rubric",
}
`;

exports[`Rubric editor plugins.renderNode calls next 1`] = `undefined`;

exports[`Rubric editor plugins.renderNode renders a modlist when passed 1`] = `
<ModList
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment.Rubric.ModList",
    }
  }
/>
`;

exports[`Rubric editor plugins.renderNode renders a modlist when passed 2`] = `
<ModList
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment.Rubric.ModList",
    }
  }
/>
`;

exports[`Rubric editor plugins.renderNode renders a modlist when passed 3`] = `
<Mod
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment.Rubric.Mod",
    }
  }
/>
`;

exports[`Rubric editor plugins.renderNode renders the rubric when passed 1`] = `
<Node
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment.Rubric",
    }
  }
/>
`;

exports[`Rubric editor slateToObo converts a Slate node to an OboNode with content 1`] = `
Object {
  "[object Object]": undefined,
  "mods": Array [
    Object {
      "attemptCondition": "mockParameter",
      "reward": "mockParameter",
    },
  ],
  "type": "pass-fail",
}
`;

exports[`Rubric editor slateToObo converts a Slate node to an OboNode with no mods 1`] = `
Object {
  "type": "pass-fail",
}
`;
